Use FunctionSchema instead of char* for dispatch

bwasti opened the pull request:
This is useful to make things run a bit faster.  It is tested (implicitly) in the diff stacked above.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27159 Use FunctionSchema instead of char* for dispatch**

Differential Revision: [D17693481](https://our.internmc.facebook.com/intern/diff/D17693481)

Commit:
Use FunctionSchema instead of char* for dispatch

ezyang reviewed (changes requested):
Based on chat, we're going to get the function schema from JIT instead

bwasti commented:
@ezyang we chatted about not using the 'try/catch' approach, and decided JIT had some tooling that could be used instead.  However, when trying to build that I figured this approach would be better.

Couple of reasons:

1) It removes the dependency on the JIT project, which has empirically been hampered by dependencies.
2) It is substantially cleaner, as using `torch/csrc/jit/operator.h` requires a refactor of headers.  I'd like to avoid that churn in the JIT.
3) JIT code in this area would need a refactored / rewritten to be used appropriately, as there are already nearly identical but (for some reason) unusable functions e.g. https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/operator.h#L188-L189

ezyang commented:
My understanding of the plan on record was that we'd refactor `operator.h` to provide the functionality we needed for this case. Quoth @zdevito:

> yeah, seems like the implementation is folded into lookupByLiteral
> not opposed to adding lookupBySchemaString or something to OperatorRegistry

I agree this is going a little off the garden path, but it's not that far off the garden path, and the refactor could have been its own patch on its own.

Re (1), I was under the impression that the big killer for JIT was when they had important files put into the mobile build; a dep from ATenDispatch.h is not that big a deal. But maybe someone on the jit side can chime in. (@suo?)

ezyang reviewed (changes requested):
It's not thread safe

Commit:
Update on "Use FunctionSchema instead of char* for dispatch"

This is useful to make things run a bit faster.  It is tested (implicitly) in the diff stacked above.


Differential Revision: [D17693481](https://our.internmc.facebook.com/intern/diff/D17693481)

bwasti commented:
thread safe now :)

ezyang commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -128,9 +129,19 @@ class CAFFE2_API ATenOpTable {
   C10_NORETURN void reportError(TensorTypeId tid) const;
 
+  const FunctionSchema& function_schema() const {
+    std::lock_guard<std::mutex> lock(mutex_);
```
It's a bit shameful we have to take out the lock every time we do an operator call.

ezyang replied:
(once flag would be more efficient here.)
